src/apps/ipsec/ipsec.lua

```diff
@@ -0,0 +1,155 @@
+module(..., package.seeall)
+local datagram = require("lib.protocol.datagram")
+local ethernet = require("lib.protocol.ethernet")
+local esp = require("lib.protocol.esp")
+local esp_tail = require("lib.protocol.esp_tail")
+local aes_128_gcm = require("apps.ipsec.aes_128_gcm")
+local lib = require("core.lib")
+local ffi = require("ffi")
+
+
+local esp_nh = 50 -- https://tools.ietf.org/html/rfc4303#section-2
+local esp_length = esp:sizeof()
+local esp_tail_length = esp_tail:sizeof()
+
+function esp_v6_new (arg)
+   local conf = arg and config.parse_app_arg(arg) or {}
+   assert(conf.mode == "aes-128-gcm", "Only supports aes-128-gcm.")
+   return { aes_128_gcm = aes_128_gcm:new(conf), seq_no = 0 }
+end
+
+
+local esp_v6_encrypt = {}
+
+function esp_v6_encrypt:new (arg)
+   local o = esp_v6_new(arg)
+   o.pad_buf = ffi.new("uint8_t[?]", o.aes_128_gcm.blocksize-1)
+   o.esp_buf = ffi.new("uint8_t[?]", o.aes_128_gcm.aad_size)
+   -- Fix me https://tools.ietf.org/html/rfc4303#section-3.3.3
```

Comment:
ESP defines an eight byte header including a four byte sequence number at the end, AES-GCM uses an eight byte initialization vector which has to be unique for each packet (they suggest a counter, e.g. sequence number). ESP talks about an ESN (extended sequence number) where only the lower four bytes of the sequence number are transmitted in the header. In that case both sender and receive have to maintain the actual counter value. Furthermore the sequence number is not allowed to wrap around. Its complicated.

For now, I just (illegally) extended the ESP header to hold an 64bit sequence number and use that as the IV. The sequence number is authenticated and so should effectively provide replay-protection.